ext/curl/curl_file.c

```diff
@@ -29,44 +29,75 @@
+static void curlfile_ctor(char *fname, size_t fname_len, char *mime, size_t mime_len, char *postname, size_t postname_len, char *buffer, size_t buffer_len, zval *return_value)
 {
-	char *fname = NULL, *mime = NULL, *postname = NULL;
-	size_t fname_len, mime_len, postname_len;
 	zval *cf = return_value;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|ss", &fname, &fname_len, &mime, &mime_len, &postname, &postname_len) == FAILURE) {
-		return;
-	}
-
 	if (fname) {
-		zend_update_property_string(curl_CURLFile_class, cf, "name", sizeof("name")-1, fname);
+		zend_update_property_stringl(curl_CURLFile_class, cf, "name", sizeof("name")-1, fname, fname_len);
 	}
 
 	if (mime) {
-		zend_update_property_string(curl_CURLFile_class, cf, "mime", sizeof("mime")-1, mime);
+		zend_update_property_stringl(curl_CURLFile_class, cf, "mime", sizeof("mime")-1, mime, mime_len);
 	}
 
 	if (postname) {
-		zend_update_property_string(curl_CURLFile_class, cf, "postname", sizeof("postname")-1, postname);
+		zend_update_property_stringl(curl_CURLFile_class, cf, "postname", sizeof("postname")-1, postname, postname_len);
+	}
+
+	if (buffer) {
+		zend_update_property_stringl(curl_CURLFile_class, cf, "buffer", sizeof("buffer")-1, buffer, buffer_len);
 	}
 }
 
-/* {{{ proto void CURLFile::__construct(string $name, [string $mimetype [, string $postfilename]])
+/* {{{ proto void CURLFile::__construct([string $name, [string $mimetype [, string $postfilename]]])
    Create the CURLFile object */
 ZEND_METHOD(CURLFile, __construct)
 {
+	char *fname = NULL, *mime = NULL, *postname = NULL, *buffer = NULL;
+	size_t fname_len, mime_len, postname_len, buffer_len;
+
 	return_value = getThis();
-	curlfile_ctor(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|sss", &fname, &fname_len, &mime, &mime_len, &postname, &postname_len) == FAILURE) {
+		return;
+	}
+
+	curlfile_ctor(fname, fname_len, mime, mime_len, postname, postname_len, buffer, buffer_len, return_value);
 }
 /* }}} */
 
 /* {{{ proto CURLFile curl_file_create(string $name, [string $mimetype [, string $postfilename]])
-   Create the CURLFile object */
+   Create the CURLFile object from file */
 PHP_FUNCTION(curl_file_create)
 {
-    object_init_ex( return_value, curl_CURLFile_class );
-    curlfile_ctor(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+	char *fname = NULL, *mime = NULL, *postname = NULL, *buffer = NULL;
+	size_t fname_len, mime_len, postname_len, buffer_len;
+
+	object_init_ex( return_value, curl_CURLFile_class );
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|ss", &fname, &fname_len, &mime, &mime_len, &postname, &postname_len) == FAILURE) {
+		return;
+	}
+
+	curlfile_ctor(fname, fname_len, mime, mime_len, postname, postname_len, buffer, buffer_len, return_value);
+}
+/* }}} */
+
+/* {{{ proto CURLFile curl_buffer_file_create(string $buffer, string $postfilename, [string $mimetype])
+   Create the CURLFile object from string buffer */
```

Comment:
Yep. We have only two variants: 
1) keep current arguments order, because for buffer we always require postname
2) keep old arguments order with "null" value $mimetype argument skip, like in mb_substr etc.: `curl_buffer_file_create($buffer, null, $postfilename)`